postgis/gserialized_supportfn.c

```diff
@@ -0,0 +1,308 @@
+	return expandfn_oid;
+}
+
+Datum postgis_index_supportfn(PG_FUNCTION_ARGS);
+PG_FUNCTION_INFO_V1(postgis_index_supportfn);
+Datum postgis_index_supportfn(PG_FUNCTION_ARGS)
+{
+	Node *rawreq = (Node *) PG_GETARG_POINTER(0);
+	Node *ret = NULL;
+
+	/*
+	* This support function is strictly for adding spatial index
+	* support.
+	*/
+	if (IsA(rawreq, SupportRequestIndexCondition))
+	{
+		/* Try to convert operator/function call to index conditions */
+		SupportRequestIndexCondition *req = (SupportRequestIndexCondition *) rawreq;
+
+		if (is_funcclause(req->node))	/* ST_Something() */
+		{
+			FuncExpr *clause = (FuncExpr *) req->node;
+			Oid funcid = clause->funcid;
+			IndexableFunction idxfn = {NULL, 0};
+			Oid opfamilyoid = req->opfamily; /* OPERATOR FAMILY of the index */
+
+			if (needsSpatialIndex(funcid, &idxfn))
+			{
+				Node *indexarg, *otherarg, *radiusarg;
+				Oid indexdatatype, otherdatatype;
+				Oid oproid;
+				int nargs = list_length(clause->args);
+
+				/*
+				* Only add an operator condition for GIST and SPGIST indexes.
+				* Effectively this means only these opclasses will get automatic
+				* indexing when used with one of the indexable functions
+				* gist_geometry_ops_2d, gist_geometry_ops_nd,
+				* spgist_geometry_ops_2d, spgist_geometry_ops_nd
+				*/
+				Oid opfamilyam = opFamilyAmOid(opfamilyoid);
+				if (opfamilyam != GIST_AM_OID &&
+				    opfamilyam != SPGIST_AM_OID &&
+				    opfamilyam != BRIN_AM_OID)
+				{
+					PG_RETURN_POINTER((Node *)NULL);
+				}
+
+				/*
+				* Need the argument types (which should always be geometry or geography
+				* since this function is only bound to those functions)
+				* to use in the operator function lookup
+				*/
+				if (nargs < 2)
+					elog(ERROR, "%s: associated with function with %d arguments", __func__, nargs);
+				if (req->indexarg > 1)
+					elog(ERROR, "%s: indexarg '%d' out of range", __func__, req->indexarg);
+
+				indexarg = req->indexarg ? lsecond(clause->args) : linitial(clause->args);
+				otherarg = req->indexarg ? linitial(clause->args) : lsecond(clause->args);
+				indexdatatype = exprType(indexarg);
+				otherdatatype = exprType(otherarg);
+
+				/*
+				* Given the index operator family and the arguments and the
+				* desired strategy number we can now lookup the operator
+				* we want (usually && or &&&).
+				*/
+				oproid = get_opfamily_member(opfamilyoid, indexdatatype, otherdatatype, idxfn.strategy_number);
+				if (oproid == InvalidOid)
+					elog(ERROR, "no spatial operator found for opfamily %u strategy %d", opfamilyoid, idxfn.strategy_number);
+
+				/*
+				* For the ST_DWithin variants we need to build a more complex return.
+				* We want to expand the non-indexed side of the call by the
+				* radius and then apply the operator.
+				* st_dwithin(g1, g2, radius) yields this, if g1 is the indexarg:
+				* g1 && st_expand(g2, radius)
```

Comment:
geography had _st_expand?

Reply:
Hrm, TBD, I see a bunch of inlined functions in geography.sql.in that I have not handled yet.

Reply:
Geography has been brought into the fold now too.